arrows/serialize/json/load_save_metadata.cxx

```diff
@@ -48,51 +51,69 @@ struct meta_item
 
     archive( CEREAL_NVP( tag ) );
 
-    // These two items are included to increase readability of the
-    // serialized form and are not used when deserializing.
-    archive(  ::cereal::make_nvp( "type", item_value.type_name() ) );
-    archive(  ::cereal::make_nvp( "name", trait.name() ) );
-
+    std::string type;
     // This is a switch on the item data type
     if ( trait.is_floating_point() )
```

Comment:
So... to update everyone else, @russelldj and I had a long offline discussion on this code. Summary: I am concerned that we need to recover the exact type, or users are likely to be confused because their `any_cast`s will unexpectedly fail. To that end, the `is_integral`, etc. methods are not useful, and in fact may be actively harmful. I believe this whole lot should be rewritten to handle many more data types (at minimum, `float`, `double` and every distinct integer type, plus e.g. strings and geo-points). However, I also want to get this PR landed, so we're thinking to address this as a follow-up.

Meanwhile, contrary to my previous comment, we *do* want to handle `int`, as that is the type used by tagged metadata items.

Reply:
How does `std::any` handle this. We should be compatible with that so it can just drop in when the time comes.

Reply:
Our `kwiver::vital::any` is designed to be compatible with the soon to be available `std::any` and if we extend our implementation, it should be compatible with the `std` version. Ideally, when we get to the point where we can use C++17 features, all we have to do is replace the type  `kwiver::vital::any` with `std::any`.

Reply:
If I understand what you're asking, this is a hack because the only floating point type we have in [metadata_tags](https://github.com/Kitware/kwiver/blob/master/vital/types/metadata_tags.h) is a double. However, it seems like from what @mwoehlke-kitware has described, if someone added a floating point type which was not a double, `any_cast` would break.

Reply:
Ahh sorry I completely missed the `std` in your comment. I can't really comment on the behavior of `std::any` but I can look into it if you'd like. Also, I'm on the same page as Matthew that this should land before we try to address any of these subtleties since there's other work depending on this code.

Reply:
@linus-sherrill, AFAIK, a) `kwiver::any` and `std::any` have the same behavior, which is that b) both `kwiver::any_cast` and `std::any_cast` perform *no* lexical conversions. That is, you have to ask for *exactly* what you put in. Even (64-bit) `long` and `long long` are not interchangeable.

As far whether `kwiver::any` and `std::any` are "interchangeable" in the behavioral sense, TTBOMK they are.

Reply:
There is an `any_converter.h` in vital/util which, I think, does lexical conversions. I don't know if that is of any use here.

Reply:
I don't think it's desired here; I think we rather want to preserve exactly the original C++ type. Conversion makes more sense at the point someone goes to try to extract a usable value from the metadata.

Reply:
There is a convention that vital metadata traits should use a limited subset of data types.  The types here should be the only ones in use.

Reply:
Is that documented somewhere? Anyway, if that's the case, I would suggest instead of `is_floating_point`, etc., having an enumeration of "allowed" types. That should allow us to avoid string comparisons (outside of `any_cast` itself, anyway) and also surprising failures if anyone violates the set of expected types.

I still submit this code ought to work (if it can, anyway; we can't do anything about someone using a custom type) if someone tries to use it on custom metadata items that don't follow the set of "allowed" types.

Reply:
Is that documented somewhere? Anyway, if that's the case, I would suggest instead of `is_floating_point`, etc., having an enumeration of "allowed" types. That should allow us to avoid string comparisons (outside of `any_cast` itself, anyway) and also surprising failures if anyone violates the set of expected types.

I still submit this code ought to work (if it can, anyway; we can't do anything about someone using a custom type) if someone tries to use it on custom metadata items that don't follow the set of "allowed" types.

Reply:
I'm not sure that it is documented, but it should be.  I would not object to formalizing this convention and possibly using an enum in a later PR, but that should not hold up this PR.  The point of the change in this PR is to add the "signed" case because were previously assuming all `int`s were unsigned, which was a problem when we added a field that needed negative integer numbers.

For the purposes of JSON we don't need or want to preserve C++ types.  The `any_converter` might be just what we need as a safety net here.  We want to just encode this data in string form.   When we read it back in we will put it back into the appropriate C++ types based on the tag, but the JSON should not encode C++ types.  In fact, it probably doesn't even need to encode "float" vs "integer" type names because I think the JSON type is supposed to be inferred from the string formatting.

Reply:
I refer you back to the first comment in this thread:

> I also want to get this PR landed, so we're thinking to address this as a follow-up.

:slightly_smiling_face:

I believe the only thing (possibly) holding this PR at present is hoping to hear back from @linus-sherrill that his concerns are adequately addressed.

> [we] were previously assuming all ints were unsigned, which was a problem when we added a field that needed negative integer numbers.

I expect it also ended badly when the `metadata` user expected the item to hold an `int` and was unable to extract the value *at all* (because it was actually a `uint64_t`). Whether or not the value was actually negative.

----

Okay, on closer inspection, it looks like we may not be able to handle custom items, period. I'll still argue (though I note it sounds like we may agree, here), that `is_signed`, etc. is perhaps not the best way to encode that an item should be one of a specific set of types. It may *work*, but it's rather obtuse.

>  The `any_converter` might be just what we need as a safety net here.

I'm not sure we need this; not in the [de]serialization, anyway. The JSON code has other ways of converting from JSON to the appropriate C++ type, and we find the appropriate type based on the tag. We just have a rather obtuse way (per previous paragraph in this comment) of extracting the appropriate type from the tag.

> the JSON should not encode C++ types

Actually, IIUC we are effectively only doing this as a convenience to the reader. Note: "These two items are included to increase readability of the serialized form and are not used when deserializing."